infoextract.js

```diff
@@ -196,50 +220,56 @@ module.exports = {
 		});
 	},
 
+	getManyPreviews(playlist) {
+		//get info for all videos in playlist, then return all non-deleted videos
+		return Promise.all(
+			playlist.map(video => video.id).map(id => this.getVideoInfo('youtube', id))
+		).then(previews => previews.filter(item => item !== undefined))
+		.catch(err => console.error('Error getting video info:', err));
+	},
+
 	getAddPreview(input) {
-		let service = this.getService(input);
+		const service = this.getService(input);
 
 		if (service !== "youtube") {
 			console.error("Unsupported input for getAddPreview");
 			throw "Unsupported input for getAddPreview";
 		}
 
-		let urlParsed = url.parse(input);
-		let queryParams = querystring.parse(urlParsed.query);
+		const urlParsed = url.parse(input);
+		const queryParams = querystring.parse(urlParsed.query);
 		if (queryParams["list"]) {
 			// there is a playlist associated with this link
 			console.log("playlist found");
 			return new Promise((resolve, reject) => {
 				this.getPlaylistYoutube(queryParams["list"]).then(playlist => {
-					let videoIds = playlist.map(item => item.id);
 					console.log(`Found ${playlist.length} videos in playlist`);
-
-					this.getVideoInfoYoutube(videoIds).then(infoResults => {
-						let addPreviewResults = [];
-						for (let i = 0; i < videoIds.length; i++) {
-							const videoInfo = infoResults[videoIds[i]];
-							if (!videoInfo) {
-								// Failed to get info for this video
-								// video has probably been deleted, skip it
-								continue;
-							}
-							let video = videoInfo;
-							if (queryParams["v"] && video.id === queryParams["v"]) {
-								video.highlight = true;
-							}
-							addPreviewResults.push(video);
-						}
-						resolve(addPreviewResults);
-					}).catch(err => {
-						console.error("Failed to compile add preview: error getting video info:", err);
-						reject(err);
-					});
+					this.getManyPreviews(playlist).then(previews => resolve(previews));
 				}).catch(err => {
 					console.error("Failed to compile add preview: error getting playlist:", err);
 					reject(err);
 				});
 			});
 		}
+		else if (urlParsed.path.startsWith('/user') || urlParsed.path.startsWith('/channel')) {
+			console.log('channel found');
+			try {
+				const channelData = {id: urlParsed.path.slice(urlParsed.path.lastIndexOf('/')+1)};
+				if (urlParsed.path.startsWith('/channel/')) {
+					channelData.channel = true;
+				}
+				else {
+					channelData.user = true;
+				}
+				return this.getChanneInfoYoutube(channelData)
+					.then(newestVideos => this.getManyPreviews(newestVideos))
+					.then(previews => previews)
```

Comment:
You aren't doing anything with this `then`, I'm pretty sure you can get rid of it.